arch/arm/include/asm/xen/hypercall.h

```diff
@@ -0,0 +1,68 @@
+/******************************************************************************
+ * hypercall.h
+ *
+ * Linux-specific hypervisor handling.
+ *
+ * Stefano Stabellini <stefano.stabellini@eu.citrix.com>, Citrix, 2012
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License version 2
+ * as published by the Free Software Foundation; or, when distributed
+ * separately from the Linux kernel or incorporated into other
+ * software packages, subject to the following license:
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this source file (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy, modify,
+ * merge, publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#ifndef _ASM_ARM_XEN_HYPERCALL_H
+#define _ASM_ARM_XEN_HYPERCALL_H
+
+#include <linux/bug.h>
+
+#include <xen/interface/xen.h>
+#include <xen/interface/platform.h>
+#include <xen/interface/xsm/flask_op.h>
+
+struct xen_dm_op_buf;
+
+long privcmd_call(unsigned call, unsigned long a1,
+		unsigned long a2, unsigned long a3,
+		unsigned long a4, unsigned long a5);
+int HYPERVISOR_xen_version(int cmd, void *arg);
+int HYPERVISOR_console_io(int cmd, int count, char *str);
+int HYPERVISOR_grant_table_op(unsigned int cmd, void *uop, unsigned int count);
+int HYPERVISOR_sched_op(int cmd, void *arg);
+int HYPERVISOR_event_channel_op(int cmd, void *arg);
+unsigned long HYPERVISOR_hvm_op(int op, void *arg);
+int HYPERVISOR_memory_op(unsigned int cmd, void *arg);
+int HYPERVISOR_physdev_op(int cmd, void *arg);
+int HYPERVISOR_vcpu_op(int cmd, int vcpuid, void *extra_args);
+int HYPERVISOR_tmem_op(void *arg);
+int HYPERVISOR_vm_assist(unsigned int cmd, unsigned int type);
+int HYPERVISOR_dm_op(domid_t domid, unsigned int nr_bufs,
+		     struct xen_dm_op_buf *bufs);
+int HYPERVISOR_platform_op_raw(void *arg);
+int HYPERVISOR_xsm_op(struct xen_flask_op *);
+
+static inline int HYPERVISOR_platform_op(struct xen_platform_op *op)
```

Comment:
What is the sense of 'static inline' in.h file ?
Really, it make a sense if function is in .c/cpp file.

Reply:
> What is the sense of 'static inline' in.h file ?

The behavior between C and C++ is quite different when defining inline helper in a header.

In the case of C, if you don't add **static**, the compiler will consider this is a prototype and ignore the body and it will result to a linking error. Below an example:

```
42sh> cat test.c
#include "test.h"

int main(void)
{
	return static_inline() + inline_only();
}
```
```
42sh> cat test.h
#ifndef __TEST_H__

inline int inline_only(void)
{
	return 1;
}

static inline int static_inline(void)
{
	return 2;
}

#endif /* !__TEST_H__ */
```

```
42sh> gcc test.c
/usr/bin/ld: /tmp/ccyxE8AD.o: in function `main':
test.c:(.text+0x1c): undefined reference to `inline_only'
collect2: error: ld returned 1 exit status
```

For C++, the compiler will just create a weak reference.

> Really, it make a sense if function is in .c/cpp file.

I would actually argue the inverse. _static inline_ only makes sense in header because the compiler will often do a better job than you to decide what to inline (_static_ function can be inlined). So unless you specify, **__attribute__((always_inline))**, your compiler is free to ignore your request to inline it.

Note that _static inline_ is still necessary in the header because without _inline_ the compiler will throw a warning if the helper is not used and compile with _-Wall_/_-Wunused-function_

Reply:
ok, thanks